### PR TITLE
sql: restore sanity in the handling of special column names

### LINKED ISSUE #16813: sql: catastrophic handling of special characters in column names (1/2)
This one, with views

```
> create table foo("x;y" int);
> create view waa as select @1 from foo;
> select * from waa;
pq: failed to parse underlying query from view "waa": syntax error at or near "y"
SELECT x;y FROM t.foo
```

cc @cockroachdb/sql-language

### LINKED ISSUE #16814: sql: catastrophic handling of special characters in column names (2/2)
```
> create table foo("x;y" int);
> create table waa as select @1 from foo;
driver: bad connection
connection lost; opening new connection: all session settings will be lost
error retrieving the transaction status: driver: bad connection
```

Server crash:
```
        panic: panic while executing "create table waa as select @1 from foo;": indexed var linked to different container (*sql.renderNode) &{planner:0xc4214681c8 source:{info:0xc420322840 plan:0xc42080a000} sourceInfo:[0xc420322840] ivarHelper:{vars:[{Idx:0 container:0xc420466400} {Idx:0 container:<nil>}] container:0xc420466400} render:[0xc421056000] columns:[{Name:@1 Typ:{} Hidden:false Omitted:false}] isStar:false numOriginalCols:1 ordering:{exactMatchCols:map[] ordering:[] unique:false} curSourceRow:[] row:[] noCopy:{}}, expected (*sql.renderNode) &{planner:0xc4214681c8 source:{info:0xc420588e40 plan:0xc421050000} sourceInfo:[0xc420588e40] ivarHelper:{vars:[{Idx:0 container:<nil>} {Idx:0 container:<nil>}] container:0xc42067ce00} render:[] columns:[] isStar:false numOriginalCols:0 ordering:{exactMatchCols:map[] ordering:[] unique:false} curSourceRow:[] row:[] noCopy:{}}

goroutine 234 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc4205a3950, 0x7f5a08d5b1c8, 0xc4204e25d0)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:180 +0xb1
panic(0x193c180, 0xc42034d840)
        /mnt/kena/go1.8/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/pgwire.(*v3Conn).serve.func1(0xc421474600, 0x7f5a08d5b1c8, 0xc420acb140)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/v3.go:360 +0xa1
panic(0x193c180, 0xc42034d840)
        /mnt/kena/go1.8/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql.(*Session).maybeRecover(0xc421468000, 0x1ae42ea, 0x9, 0xc42146b2ca, 0x27)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/session.go:1072 +0x298
panic(0x188b6c0, 0xc4201e7370)
        /mnt/kena/go1.8/src/runtime/panic.go:489 +0x2cf
github.com/cockroachdb/cockroach/pkg/sql/parser.(*IndexedVarHelper).AssertSameContainer(0xc4214699a0, 0xc4202f4ea0)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/indexed_vars.go:139 +0x18c
github.com/cockroachdb/cockroach/pkg/sql.(*nameResolutionVisitor).VisitPre(0xc421469960, 0x27c8100, 0xc4202f4ea0, 0x10, 0x188ac00, 0xc4202f4e01)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/select_name_resolution.go:74 +0x3b7
github.com/cockroachdb/cockroach/pkg/sql/parser.WalkExpr(0x27b6c80, 0xc421469960, 0x27c8100, 0xc4202f4ea0, 0x1, 0x1, 0x0)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/parser/walk.go:555 +0x4c
github.com/cockroachdb/cockroach/pkg/sql.(*planner).resolveNames(0xc4214681c8, 0x27c8100, 0xc4202f4ea0, 0xc4203f0138, 0x1, 0x1, 0xc42177f170, 0x2, 0x2, 0x27bfe40, ...)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/select_name_resolution.go:201 +0x205
github.com/cockroachdb/cockroach/pkg/sql.(*planner).analyzeExpr(0xc4214681c8, 0x7f5a08d5b1c8, 0xc420395500, 0x27c8100, 0xc4202f4ea0, 0xc4203f0138, 0x1, 0x1, 0xc42177f170, 0x2, ...)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/analyze.go:1654 +0x28d
github.com/cockroachdb/cockroach/pkg/sql.(*planner).computeRender(0xc4214681c8, 0x7f5a08d5b1c8, 0xc420395500, 0x27c8100, 0xc4202f4ea0, 0x0, 0x0, 0x27cef80, 0x2c862c0, 0xc4203f0138, ...)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/targets.go:48 +0x164
github.com/cockroachdb/cockroach/pkg/sql.(*planner).computeRenderAllowingStars(0xc4214681c8, 0x7f5a08d5b1c8, 0xc420395500, 0x27c8100, 0xc4202f4ea0, 0x0, 0x0, 0x27cef80, 0x2c862c0, 0xc4203f0138, ...)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/targets.go:78 +0x21c
github.com/cockroachdb/cockroach/pkg/sql.(*renderNode).initTargets(0xc42067ce00, 0x7f5a08d5b1c8, 0xc420395500, 0xc4202f4ee0, 0x1, 0x1, 0xc42034d7c0, 0x2, 0x2, 0x1, ...)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/render.go:338 +0x29a
github.com/cockroachdb/cockroach/pkg/sql.(*planner).SelectClause(0xc4214681c8, 0x7f5a08d5b1c8, 0xc420395500, 0xc420162bd0, 0x0, 0x0, 0x0, 0x0, 0xc42034d7c0, 0x2, ...)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/render.go:233 +0x2dc
github.com/cockroachdb/cockroach/pkg/sql.(*planner).newPlan(0xc4214681c8, 0x7f5a08d5b1c8, 0xc420395500, 0x27c8540, 0xc420162bd0, 0xc42034d7c0, 0x2, 0x2, 0x2, 0xc42034d620, ...)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:325 +0x105a
github.com/cockroachdb/cockroach/pkg/sql.(*planner).Insert(0xc4214681c8, 0x7f5a08d5b1c8, 0xc420395500, 0xc4213cddb0, 0x0, 0x0, 0x0, 0x19a1c00, 0xc421057b60, 0x0, ...)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/insert.go:122 +0x4ea
github.com/cockroachdb/cockroach/pkg/sql.(*createTableNode).Start(0xc42162fef0, 0x7f5a08d5b1c8, 0xc420395500, 0x0, 0x0)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/create.go:688 +0xae1
github.com/cockroachdb/cockroach/pkg/sql.(*planner).startPlan(0xc4214681c8, 0x7f5a08d5b1c8, 0xc420395500, 0x27c6c80, 0xc42162fef0, 0x0, 0xffffffee)
        /mnt/kena/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:208 +0x8e
```

### PR DESCRIPTION
**Pull request by knz**
Fixes  #16813.
Fixes  #16814.

This should make it into the next 1.0.x, do you agree?

### CODE REVIEW

**Comment by knz** [comment]
tests fixed, ptal

**Comment by jordanlewis** [comment]
:lgtm: but I'd feel better if @raduberinde also had a look before merging.

---

Reviewed 1 of 1 files at r1, 8 of 8 files at r2, 5 of 5 files at r3.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-:-KnuzLHmRo6SgQfQVaGL:bobh8e8)*

File: pkg/sql/parser/indexed_vars.go [hunk 1]
```diff
@@ -238,3 +254,19 @@ func (h *IndexedVarHelper) VisitPre(expr Expr) (recurse bool, newExpr Expr) {
 
 // VisitPost implements the Visitor interface.
 func (*IndexedVarHelper) VisitPost(expr Expr) Expr { return expr }
+
+type unboundContainerType struct{}
+
+var unboundContainer = &unboundContainerType{}
```
Comment (RaduBerinde) [inline pkg/sql/parser/indexed_vars.go]: Please add a comment explaining when this is used, and how it differs from a `nil` container. I think the gist of it is that it indicates that we were previously bound to a container but we aren't anymore, right?

File: pkg/sql/parser/indexed_vars.go [hunk 2]
```diff
@@ -111,35 +122,42 @@ type IndexedVarHelper struct {
...
-// BindIfUnbound attaches an IndexedVar to an existing container.
-// This is needed for standalone column ordinals created during parsing.
-func (h *IndexedVarHelper) BindIfUnbound(ivar *IndexedVar) error {
-	if ivar.container != nil {
-		return nil
-	}
+// BindIfUnbound ensures the IndexedVar is attached to this helper's container.
+// - for freshly created IndexedVars (with a nil container) this will bind in-place.
+// - for already bound IndexedVar, bound to this container, this will return the same ivar unchanged.
+// - for ordinal references (with an explicit unboundedContainer) this will return a new var.
```
Comment (RaduBerinde) [inline pkg/sql/parser/indexed_vars.go]: unboundContainer

**Review by RaduBerinde — COMMENTED** [review]
LGTM, some minor comments.
I do think it should make it in a 1.0.x release, but I am a bit concerned at the amount of code touched. Maybe we want it on master for a while before doing that.

**Comment by bdarnell** [comment]
👍 for inclusion in 1.0.4. 1.0.3 is going out this week, so we'll have this on master for a couple of weeks before we cut the next release.

**Comment by tamird** [comment]
> This it turns fixes a panic inducing bug in the handling of CREATE
> TABLE AS, when the source select contained ordinal references.

s/it turns/in turn/

Also, can you put each "Fixes #xxx" in the commit that fixes it? Looks like each issue is fixed by one commit, but please correct me if I'm wrong.

---

Reviewed 1 of 1 files at r1, 8 of 8 files at r2, 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/window.go, line 402 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-KntR20Z7n3OhClDjjG3-r3-402:-Ko9u-ap8I4X79ul3V5T:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/window.go#L402)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Do we want to check this assumption? (panic if we return anything other than `ivar`)
</blockquote></details>

+1

---

*[pkg/sql/logictest/testdata/logic_test/name_escapes, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-Ko9sPfa6lz75ss6z7ok:-Ko9sPfa6lz75ss6z7ol:b-6kfxfe) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b25008a2c76fb7bb42a54f6c2e5b1b8b6c1c960/pkg/sql/logictest/testdata/logic_test/name_escapes#L41)):*
> ```
> CREATE TABLE unused (x ")
> ) INTERLEAVE IN PARENT "woo; DROP USER humpty;" (x)
> ```

errant empty newline

looks like r2 "fixed" it

---

*[pkg/sql/parser/indexed_vars.go, line 107 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-Ko9uf6VLiKi-zC-bB7t:-Ko9uf6VLiKi-zC-bB7u:bm0wn7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/parser/indexed_vars.go#L107)):*
> ```Go
> // NewIndexedVar is a helper routine to create a standalone Indexedvar
> // with the given index value. This needs to undergo BindIfUnbound()
> // below before it can be fully used.  The difference with ordinal
> ```

double space

---

*[pkg/sql/parser/indexed_vars.go, line 109 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-Ko9v-bO_7jeBbrqVxaf:-Ko9v-bO_7jeBbrqVxag:b-1ldjzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/parser/indexed_vars.go#L109)):*
> Do not use NewIndexedVar for AST nodes

Consider giving this its own paragraph.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-:-Ko9tbexbQnmJPkbwoqd:b8yhgdj)*

**Comment by knz** [comment]
I made the proposed changes.

I don't usually indicate which commit fixes which issue. Instead I explain (and repeat) what the issue is in the commit message that fixes each issue. This ensures that the git history can be read and understood without the github context. The relationship between PRs and issues is a github thing. 

---

Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/sql/window.go, line 402 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-KntR20Z7n3OhClDjjG3-r3-402:-KoDTuL3dfrB2A0CLjVu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/window.go#L402)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/expr.go, line 44 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-KntR20Z7n3OhClDjjG0-r3-44:-KoDTwDsT-Nq2M4R-Z6S:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/distsqlrun/expr.go#L44)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think we should return `false, expr` in the error case, or we may screw up the expression.
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/name_escapes, line 41 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-Ko9sPfa6lz75ss6z7ok:-KoDUWxdH6mtihpFPkg3:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/3b25008a2c76fb7bb42a54f6c2e5b1b8b6c1c960/pkg/sql/logictest/testdata/logic_test/name_escapes#L41)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

errant empty newline

looks like r2 "fixed" it
</blockquote></details>

Fixed.

---

*[pkg/sql/parser/indexed_vars.go, line 107 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-Ko9uf6VLiKi-zC-bB7t:-KoDUgEhR8K-ogdUEFvG:b-7ql1td) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/parser/indexed_vars.go#L107)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

double space
</blockquote></details>

fixed

---

*[pkg/sql/parser/indexed_vars.go, line 109 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-Ko9v-bO_7jeBbrqVxaf:-KoDUhejdFhN9Kb1iSK-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/parser/indexed_vars.go#L109)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> Do not use NewIndexedVar for AST nodes

Consider giving this its own paragraph.
</blockquote></details>

Done.

---

*[pkg/sql/parser/indexed_vars.go, line 128 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-KntR20Z7n3OhClDjjG1-r3-128:-KoDUkK6vxuaTiwURqPh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/parser/indexed_vars.go#L128)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

unboundContainer
</blockquote></details>

Done.

---

*[pkg/sql/parser/indexed_vars.go, line 260 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-KntR20Z7n3OhClDjjG1-r3-260:-KoDVAmysCjzbUrFdX24:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/parser/indexed_vars.go#L260)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Please add a comment explaining when this is used, and how it differs from a `nil` container. I think the gist of it is that it indicates that we were previously bound to a container but we aren't anymore, right?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-:-KoDVMzAvOJVWd9GNf3k:b-mkcdoc)*

**Comment by tamird** [comment]
> I don't usually indicate which commit fixes which issue. Instead I explain (and repeat) what the issue is in the commit message that fixes each issue. This ensures that the git history can be read and understood without the github context. The relationship between PRs and issues is a github thing.

I agree that explaining the issue in the commit message is helpful, but what's the downside of _also_ including the issue reference? Yes, it's a github thing, but since the atomic unit of change that fixes the issue is the commit, it seems strictly beneficial to include.

---

Reviewed 12 of 12 files at r4, 8 of 8 files at r5, 5 of 5 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[pkg/sql/window.go, line 402 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-KntR20Z7n3OhClDjjG3-r3-402:-KoIM4VH418p4dPye-YB:b-ynzdvy) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/window.go#L402)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

The comment above should be updated now; also, if you don't mind splitting the two cases:

```
			if newIvar, err := aggHelper.BindIfUnbound(ivar); err != nil {
				panic(err)
			} else if newIvar != ivar {
				panic(fmt.Sprintf("unexpected binding: %v, expected: %v", newIvar, ivar"))
			}
```

---

*[pkg/sql/parser/indexed_vars.go, line 111 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-KoIMZ_JEGrpSCJB9ObH:-KoIMZ_JEGrpSCJB9ObI:b-xzd9xf) ([raw file](https://github.com/cockroachdb/cockroach/blob/1752e18a929aff0e2786436c28d9db788b9b2e90/pkg/sql/parser/indexed_vars.go#L111)):*
> ```Go
> // in-place by BindIfUnbound.
> //
> // Do not use NewIndexedVar for AST nodes
> ```

re-wrap this plz

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-:-KoILeBqZZVTNVycnE15:b3ces3u)*

**Comment by knz** [comment]
okay, that works for me; I changed the commit messages.

---

Review status: 0 of 12 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/window.go, line 402 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-KntR20Z7n3OhClDjjG3-r3-402:-KoJPq2mSjY0Jy32IZOL:bt4ox15) ([raw file](https://github.com/cockroachdb/cockroach/blob/56e8fbd4fce3555bc28b4e9fedc2ed175b28d2b7/pkg/sql/window.go#L402)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

The comment above should be updated now; also, if you don't mind splitting the two cases:

```
			if newIvar, err := aggHelper.BindIfUnbound(ivar); err != nil {
				panic(err)
			} else if newIvar != ivar {
				panic(fmt.Sprintf("unexpected binding: %v, expected: %v", newIvar, ivar"))
			}
```
</blockquote></details>

Done. Thanks for the suggestion!

---

*[pkg/sql/parser/indexed_vars.go, line 111 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-KoIMZ_JEGrpSCJB9ObH:-KoJPerPLQ_OAuYMP1cA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1752e18a929aff0e2786436c28d9db788b9b2e90/pkg/sql/parser/indexed_vars.go#L111)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

re-wrap this plz
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-:-KoJP5DmgSZ0PcAeGStX:bog1gp5)*

**Comment by tamird** [comment]
:lgtm: thanks!

---

Reviewed 7 of 8 files at r8, 5 of 5 files at r9.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16815#-:-KoJQiJ9IxyxwzZumzFt:brfqlux)*

**Comment by knz** [comment]
TFYRs!

**Pull request merged**